File: domain/src/main/java/com/github/anrimian/musicplayer/domain/repositories/EditorRepository.java
package com.github.anrimian.musicplayer.domain.repositories;

import com.github.anrimian.musicplayer.domain.models.composition.FullComposition;
import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource;
import com.github.anrimian.musicplayer.domain.models.folders.FileSource;
import com.github.anrimian.musicplayer.domain.models.genres.ShortGenre;
import com.github.anrimian.musicplayer.domain.models.image.ImageSource;

import java.util.Collection;
import java.util.List;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;

public interface EditorRepository {

    Completable changeCompositionGenre(long compositionId,
                                       CompositionContentSource source,
                                       ShortGenre oldGenre,
                                       String newGenre);

    Completable addCompositionGenre(long compositionId,
                                    CompositionContentSource source,
                                    String newGenre);

    Completable removeCompositionGenre(long compositionId,
                                       CompositionContentSource source,
                                       ShortGenre genre);

    Completable changeCompositionAuthor(long compositionId,
                                        CompositionContentSource source,
                                        String newAuthor);

    Completable changeCompositionAlbumArtist(long compositionId,
                                             CompositionContentSource source,
                                             String newAuthor);

    Completable changeCompositionAlbum(long compositionId,
                                       CompositionContentSource source,
                                       String newAlbum);

    Completable changeCompositionTitle(long compositionId, CompositionContentSource source, String title);

    Completable changeCompositionLyrics(long compositionId,
                                        CompositionContentSource source,
                                        String text);

    Completable changeCompositionFileName(FullComposition composition, String fileName);

    Completable changeFolderName(long folderId, String folderName);

    Completable moveFiles(Collection<FileSource> files,
                          @Nullable Long fromFolderId,
                          @Nullable Long toFolderId);

    Completable moveFilesToNewDirectory(Collection<FileSource> files,
                                        @Nullable Long fromFolderId,
                                        @Nullable Long targetParentFolderId,
                                        String directoryName);

    Completable updateAlbumName(String name,
                                List<Long> compositionIds,
                                List<CompositionContentSource> sources,
                                long albumId,
                                BehaviorSubject<Long> editingSubject);

    Completable updateAlbumArtist(String artist,
                                  List<Long> compositionIds,
                                  List<CompositionContentSource> sources,
                                  long albumId,
                                  BehaviorSubject<Long> editingSubject);

    Completable updateArtistName(String name,
                                 List<Long> compositionIds,
                                 List<CompositionContentSource> sources,
                                 long artistId,
                                 BehaviorSubject<Long> editingSubject);

    Completable updateGenreName(String name,
                                List<Long> compositionIds,
                                List<CompositionContentSource> sources,
                                long genreId,
                                BehaviorSubject<Long> editingSubject);

    Completable changeCompositionAlbumArt(long compositionId,
                                          CompositionContentSource source,
                                          ImageSource imageSource);

    Completable removeCompositionAlbumArt(long compositionId, CompositionContentSource source);

    Completable updateTagsFromSource(CompositionContentSource source, FullComposition fullComposition);
}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/repositories/EditorRepository.kt
package com.github.anrimian.musicplayer.domain.repositories

import com.github.anrimian.musicplayer.domain.models.composition.FullComposition
import com.github.anrimian.musicplayer.domain.models.composition.change.ChangedCompositionPath
import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.folders.FileSource
import com.github.anrimian.musicplayer.domain.models.image.ImageSource
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject

interface EditorRepository {
    
    fun changeCompositionGenre(
        compositionId: Long,
        source: CompositionContentSource,
        oldGenre: String,
        newGenre: String
    ): Completable

    fun addCompositionGenre(
        compositionId: Long,
        source: CompositionContentSource,
        newGenre: String
    ): Completable

    fun moveGenre(
        compositionId: Long,
        source: CompositionContentSource,
        from: Int,
        to: Int
    ): Completable

    fun removeCompositionGenre(
        compositionId: Long,
        source: CompositionContentSource,
        genre: String
    ): Completable

    fun restoreRemovedCompositionGenre(): Completable

    fun changeCompositionAuthor(
        compositionId: Long,
        source: CompositionContentSource,
        newAuthor: String?
    ): Completable

    fun changeCompositionAlbumArtist(
        compositionId: Long,
        source: CompositionContentSource,
        newAuthor: String?
    ): Completable

    fun changeCompositionAlbum(
        compositionId: Long,
        source: CompositionContentSource,
        newAlbum: String?
    ): Completable

    fun changeCompositionTitle(
        compositionId: Long,
        source: CompositionContentSource,
        title: String
    ): Completable

    fun changeCompositionTrackNumber(
        compositionId: Long,
        source: CompositionContentSource,
        trackNumber: Long?
    ): Completable

    fun changeCompositionDiscNumber(
        compositionId: Long,
        source: CompositionContentSource,
        discNumber: Long?
    ): Completable

    fun changeCompositionComment(
        compositionId: Long,
        source: CompositionContentSource,
        text: String?
    ): Completable

    fun changeCompositionLyrics(
        compositionId: Long,
        source: CompositionContentSource,
        text: String?
    ): Completable

    fun changeCompositionFileName(
        compositionId: Long,
        fileName: String
    ): Single<ChangedCompositionPath>

    fun changeFolderName(
        folderId: Long,
        newFolderName: String
    ): Single<List<ChangedCompositionPath>>

    fun moveFiles(
        files: Collection<FileSource>,
        fromFolderId: Long?,
        toFolderId: Long?
    ): Single<List<ChangedCompositionPath>>

    fun moveFilesToNewDirectory(
        files: Collection<FileSource>,
        fromFolderId: Long?,
        targetParentFolderId: Long?,
        directoryName: String?
    ): Single<List<ChangedCompositionPath>>

    fun updateAlbumName(
        name: String?,
        compositionIds: List<Long>,
        sources: List<CompositionContentSource>,
        albumId: Long,
        editingSubject: BehaviorSubject<Long>
    ): Completable

    fun updateAlbumArtist(
        artist: String?,
        compositionIds: List<Long>,
        sources: List<CompositionContentSource>,
        albumId: Long,
        editingSubject: BehaviorSubject<Long>
    ): Completable

    fun updateArtistName(
        name: String?,
        compositionIds: List<Long>,
        sources: List<CompositionContentSource>,
        artistId: Long,
        editingSubject: BehaviorSubject<Long>
    ): Completable

    fun updateGenreName(
        name: String,
        compositionIds: List<Long>,
        sources: List<CompositionContentSource>,
        genreId: Long,
        editingSubject: BehaviorSubject<Long>
    ): Completable

    fun changeCompositionAlbumArt(
        compositionId: Long,
        source: CompositionContentSource,
        imageSource: ImageSource?
    ): Completable

    fun removeCompositionAlbumArt(
        compositionId: Long,
        source: CompositionContentSource
    ): Completable

    fun updateTagsFromSource(
        source: CompositionContentSource,
        composition: FullComposition
    ): Completable

}